parlai/scripts/torchscript.py

```diff
@@ -0,0 +1,751 @@
+                    bigram = pair
+            if '\n'.join(bigram) not in self.bpe_ranks:
+                break
+            first, second = bigram
+            new_word: List[str] = []
+            i = 0
+            while i < len(word):
+                found = False
+                for j in range(i, len(word)):
+                    if word[j] == first:
+                        new_word.extend(word[i:j])
+                        i = j
+                        found = True
+                        break
+                if not found:
+                    new_word.extend(word[i:])
+                    break
+
+                if word[i] == first and i < len(word) - 1 and word[i + 1] == second:
+                    new_word.append(first + second)
+                    i += 2
+                else:
+                    new_word.append(word[i])
+                    i += 1
+            word = new_word.copy()
+            if len(word) == 1:
+                break
+            else:
+                pairs = self.get_pairs(word)
+        return word
+
+    def helper_encode(self, text: str) -> List[str]:
+        """
+        Tokenize text.
+
+        :param text:
+            text to tokenize
+
+        :return tokens:
+            A list of tokens
+        """
+        bpe_tokens: List[str] = []
+        for token in self.findall(text):
+            byte_encoded: List[str] = []
+            for b in token:
+                byte_encoded.append(self.byte_encoder[ord(b)])
+            encoded: List[str] = []
+            for bpe_token in self.bpe(byte_encoded):
+                encoded.append(self.encoder[bpe_token])
+            bpe_tokens.extend(encoded)
+        return bpe_tokens
+
+    def decode(self, tokens: List[str]) -> str:
+        """
+        Decode list of tokens into a text string.
+
+        :param tokens:
+            list of tokens
+
+        :return text:
+            decoded text
+        """
+        text = self.helper_decode(tokens)
+        if self.add_prefix_space:
+            assert text.startswith(' ')
+            text = text.lstrip(' ')
+        return text
+
+    def helper_decode(self, tokens: List[str]) -> str:
+        """
+        Decode list of tokens into text string.
+
+        :param tokens:
+            list of tokens
+
+        :return:
+            decoded text
+        """
+        chars: List[str] = []
+        for token in tokens:
+            decoded_token = self.decoder[token]
+            token_chars = self.utf8_chars(decoded_token)
+            for char in token_chars:
+                if not torch.jit.is_scripting():
+                    # We iterate over "char", which is supposed to be a single
+                    # character, because the TorchScripted version of the code
+                    # correctly splits a string into single characters in
+                    # self.utf8_chars() but the non-TorchScripted version doesn't
+                    chars.extend([real_char for real_char in char])
```

Comment:
```suggestion
                    chars.extend(list(real_char for real_char in char]))
```

Reply:
@kauterry what's the rationale behind this suggestion? curious if it's better time/space complexity or something

Reply:
My bad, I actually meant to suggest `chars.extend(list(char))`.
It's a style-based suggestion as you can see here:
<img width="842" alt="Screen Shot 2021-03-17 at 2 04 33 PM" src="https://user-images.githubusercontent.com/29103305/111538428-e7f60100-8729-11eb-9098-162cc8c238fe.png">
I guess we can skip this since @EricMichaelSmith says TorchScript throws an error with `list()`.

Reply:
the real suggestion is `chars.extend(list(char))`

Reply:
great, thanks for the clarification makes sense!

Reply:
Just checked - actually, `list(char)` is fine, it's just the generator comprehension `(real_char for real_char in char)` that had an issue. Changed to `list(char)`